### ISSUE TITLE
Unable to perform a multipart PUT

### ISSUE THREAD

**Issue by rkoster** [typhoeus/typhoeus#389]
For some reason uploading a file via a PUT request does not seem to work.

Http post works as expected:

``` ruby
Typhoeus.post("example.com/upload",
  body: {
    foo: [],
    file: File.open(upload, "r")
  },
  headers: { Authorization: "token" }
)
```

result:

```
POST /upload HTTP/1.1
Host: example.com
Accept: */*
User-Agent: Typhoeus - https://github.com/typhoeus/typhoeus
Authorization: token
Content-Length: 662
Expect: 100-continue
Content-Type: multipart/form-data; boundary=----------------------------62c3887bcd8c
```

While http put does not:

``` ruby
Typhoeus.put("example.com/upload",
  body: {
    foo: [],
    file: File.open(upload, "r")
  },
  headers: { Authorization: "token" }
)
```

result:

```
PUT /upload HTTP/1.1
Host: example.com
Accept: */*
User-Agent: Typhoeus - https://github.com/typhoeus/typhoeus
Authorization: token
Content-Length: 120
Expect: 100-continue
```

As can be seen above the `Content-type: multipart/from-data` and `boundary` headers are missing.

**Comment by nick-desteffen** [comment]
We are having issues with this as well.  Even when we set the content type to `multipart/form-data` the file field we are sending across just comes across as an array:

This is what we are calling:
```ruby
Typhoeus.put('http://example.com', 
  headers: {
    "content-type"=>"multipart/form-data"
  }, 
  body: {
    article: {
      title: "test", 
      body:  "<p>test</p>\r\n", 
      banner_image: #<File:/tmp/unnamed.jpg>
    }
  }
)
```
On the other side this is what we are getting:

```
"banner_image"=> "[\"unnamed.jpg\", \"image/jpeg\", \"/tmp/unnamed.jpg\"]"
```

I'll see about setting up a simple client / server repo that can reproduce the problem.

**Comment by nick-desteffen** [comment]
Ok, here is a repo that demonstrates the issue:
https://github.com/nick-desteffen/typhoeus-put-file

**Comment by i0rek** [comment]
@nick-desteffen thanks for reporting! I am sorry you are having trouble with that. I don't have much time these days and it would be very much appreciated if anyone wants to take a stab at it.

**Comment by nick-desteffen** [comment]
I suspect the issue is in the ethon gem, I'm not very familiar with it though.  I can take a stab at it sometime over the next week.

**Comment by i0rek** [comment]
yes. you can start digging here: https://github.com/typhoeus/ethon/blob/e2300cb201e3374c8cffca1b272d62dbda4f8d41/lib/ethon/easy/queryable.rb#L77-L84.

**Comment by nick-desteffen** [comment]
Ok, I did some digging, I don't think the `queryable#file_info` is the problem, it works the same for both PUT and POST requests.

I'm not sure why these are so different:
https://github.com/typhoeus/ethon/blob/master/lib/ethon/easy/http/putable.rb#L14-L20
https://github.com/typhoeus/ethon/blob/master/lib/ethon/easy/http/postable.rb#L15-L27

**Comment by i0rek** [comment]
@nick-desteffen I think you found an interesting piece of code. thats the implementation for put and post according to the libcurl docs: http://curl.haxx.se/libcurl/c/CURLOPT_UPLOAD.html and http://curl.haxx.se/libcurl/c/CURLOPT_POST.html.

**Comment by lucaspiller** [comment]
Any updates on this? I want to switch to typheous/excon, but this is a blocker for our use case.

**Comment by nick-desteffen** [comment]
AFAIK the problem still exists.  We are using Typhoeus to submit to a Rails application, so we just submitted the request as a `POST` with the `method` parameter as `PUT` for a workaround since Rails will take that into consideration for the routing.

**Comment by garygao** [comment]
Hi @i0rek, is there have any updates on this? Or any alternative solution? I also have the same issue now. Thanks!

**Comment by pmvilaca** [comment]
Hi,

I am also facing this problem.

**Comment by pmvilaca** [comment]
> AFAIK the problem still exists. We are using Typhoeus to submit to a Rails application, so we just submitted the request as a POST with the method parameter as PUT for a workaround since Rails will take that into consideration for the routing.

@nick-desteffen how do you fake the request? I'm facing exactly the same issue, but I'm not finding a way to use that workaround

**Comment by nick-desteffen** [comment]
If you are using Ruby on Rails, you can override the HTTP method in it's router.  

If you do this:
```ruby
params = {
  foo: 'bar'
}
params[:_method] = 'PUT'
Typhoeus.post(url, body: params)
```

And have a route defined against a PUT method, it will get routed correctly.  It is an old hack to support legacy browsers that don't support anything other than POST and GET requests.

**Comment by nick-desteffen** [comment]
For more info, see: http://guides.rubyonrails.org/form_helpers.html#how-do-forms-with-patch-put-or-delete-methods-work-questionmark

**Comment by pmvilaca** [comment]
Thank you very much @nick-desteffen 

It worked as expected

**Comment by alexanderadam** [comment]
Just to be sure: since 2014 there's no bugfix or workaround for this, right?
Or does anyone have a hint for non-Rails APIs?

**Comment by ndbroadbent** [comment]
Hi all, I've just submitted a PR that fixes this issue in https://github.com/typhoeus/ethon/pull/245

**Comment by felipedmesquita** [comment]
Hello everyone!
Thanks, @ndbroadbent, for your PR in Ethon. I'm studying how to best approach this to ensure we avoid backwards incompatible changes as much as possible.
### Here is my current understanding of the issue:
Ethon, being a libcurl wrapper, has different behavior for POST and PUT requests. This seems to originate from libcurl, where PUT requests are intended for file uploads so it does not insert headers like content-type nor messes with the body formatting [0]. 
Calling `Typhoeus.put` with a file in the body to make a multipart formpost does not go through the POST machinery that would add the headers and format the request body appropriately, causing the upload to fail if sent to a sever expecting a form submission format.
The Everything Curl book's page on PUT requests using the curl cli [1] suggests using the POST options to build the request and overriding the method to PUT if formdata formatting is desired:

> The similarities between POST and PUT also allows you to send a PUT with a string by using the regular curl POST mechanism using -d but asking for it to use a PUT instead:
> `curl -d "data to PUT" -X PUT http://example.com/new/resource/file`

The `-X` option for curl [2] overrides just the http method without modifying any other behavior in building and executing the request, which seems to be what we want here. We can use libcurl's equivalent option, `CURLOPT_CUSTOMREQUEST` [3], in Typhoeus to make PUT requests that have the same behavior of POSTs:

```ruby
require "typhoeus"
require "tempfile"
require "json"

Tempfile.open(["demo", ".txt"]) do |file|
  file.write("Hello from Typhoeus!")
  file.rewind

  body = {
    file: file,
    note: "just some text"
  }

  request = Typhoeus::Request.new(
    "https://httpbin.org/anything",
    method: :post,
    body: body,
    customrequest: "put"
  )

  response = request.run

  data = JSON.parse(response.body)
  puts "HTTP method:            #{data["method"]}"
  puts "Content-Type header:    #{data["headers"]["Content-Type"]}"
  puts "Form data fields:       #{data["form"].inspect}"
  puts "Files data:             #{data["files"].inspect}"
end
```
Outputs:
```
HTTP method:            PUT
Content-Type header:    multipart/form-data; boundary=------------------------Mh2oc2uvGDtUbuuVZmr0t8
Form data fields:       {"note" => "just some text"}
Files data:             {"file" => "Hello from Typhoeus!"}
```
This works with rails without needing to include the `_method` param. I've published an example app [4] of this working with the default scaffold controller.

I'm unsure how and if we should deviate from the behavior of libcurl and curl cli on this case. I think the `customrequest` workaround might be simple enough to implement in the application layer, but would appreciate the thoughts of everyone here.

[0] http://everything.curl.dev/libcurl-http/upload.html#http-put
[1] https://everything.curl.dev/http/put.html
[2] https://curl.se/docs/manpage.html#-X
[3] https://curl.se/libcurl/c/CURLOPT_CUSTOMREQUEST.html
[4] https://github.com/felipedmesquita/test_multipart_formpost
